packages/asset-swapper/src/quote_consumers/exchange_swap_quote_consumer.ts

```diff
@@ -69,62 +70,58 @@ export class ExchangeSwapQuoteConsumer implements SwapQuoteConsumerBase<Exchange
 
     public async getSmartContractParamsOrThrowAsync(
         quote: SwapQuote,
-        _opts: Partial<SwapQuoteGetOutputOptsBase>,
+        _opts: Partial<SwapQuoteGetOutputOpts> = {},
     ): Promise<SmartContractParamsInfo<ExchangeSmartContractParams>> {
         assert.isValidSwapQuote('quote', quote);
 
         const { orders } = quote;
 
         const signatures = _.map(orders, o => o.signature);
 
-        const optimizedOrders = swapQuoteConsumerUtils.optimizeOrdersForMarketExchangeOperation(orders, quote.type);
-
         let params: ExchangeSmartContractParams;
         let methodName: string;
 
         if (quote.type === MarketOperation.Buy) {
             const { makerAssetFillAmount } = quote;
 
             params = {
-                orders: optimizedOrders,
+                orders,
                 signatures,
                 makerAssetFillAmount,
                 type: MarketOperation.Buy,
             };
 
-            methodName = 'marketBuyOrders';
+            methodName = 'marketBuyOrdersFillOrKill';
         } else {
             const { takerAssetFillAmount } = quote;
 
             params = {
-                orders: optimizedOrders,
+                orders,
                 signatures,
                 takerAssetFillAmount,
                 type: MarketOperation.Sell,
             };
 
-            methodName = 'marketSellOrders';
+            methodName = 'marketSellOrdersFillOrKill';
         }
 
-        const methodAbi = utils.getMethodAbiFromContractAbi(
-            this._contractWrappers.exchange.abi,
-            methodName,
-        ) as MethodAbi;
+        const methodAbi = utils.getMethodAbiFromContractAbi(this._exchangeContract.abi, methodName) as MethodAbi;
 
         return {
             params,
-            toAddress: this._contractWrappers.exchange.address,
+            toAddress: this._exchangeContract.address,
             methodAbi,
+            ethAmount: quote.worstCaseQuoteInfo.protocolFeeInEthAmount,
```

Comment:
Added to pay protocol fees.

Reply:
nice catch, I completely forgot about adding that in.